src/ssl.c

```diff
@@ -48370,6 +48370,357 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
+                remainLen -= lineLen;
+                lineLen = wolfSSL_BIO_gets(in, &section[sectionLen], remainLen);
+                if (lineLen<=0) return NULL;
+            }
+
+            if (boundary) XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+        }
+    }
+    else if (curHeader && (!XSTRNCMP(curHeader->body, "application/pkcs7-mime", XSTRLEN("application/pkcs7-mime")) ||
+                           !XSTRNCMP(curHeader->body, "application/x-pkcs7-mime", XSTRLEN("application/x-pkcs7-mime")))) {
+        sectionLen = wolfSSL_BIO_get_len(in);
+        if (sectionLen <= 0) return NULL;
+        section = XMALLOC((sectionLen+1)*sizeof(char), NULL, DYNAMIC_TYPE_PKCS7);
+        if (!section) return NULL;
+        ret = wolfSSL_BIO_read(in, section, sectionLen);
+        if (ret < 0 || ret != sectionLen) {
+            WOLFSSL_MSG("Error reading input BIO.\n");
+            return NULL;
+        }
+    }
+    else {
+        WOLFSSL_MSG("S/MIME headers not found.\n");
+        return NULL;
+    }
+
+    curHeader = MIME_find_header_name("Content-Transfer-Encoding", allHeaders);
+    if (!curHeader) WOLFSSL_MSG("Content-Transfer-Encoding header not found, assuming base64 encoding.");
+    else if (XSTRNCMP(curHeader->body, "base64", XSTRLEN("base64"))) {
+        WOLFSSL_MSG("S/MIME encodings other than base64 not currently supported.\n");
+        return NULL;
+    }
+
+    if (!section || sectionLen <= 0) return NULL;
+    /* Strip trailing newlines. */
+    while ((section[sectionLen-1] == '\r' || section[sectionLen-1] == '\n') && sectionLen > 0) sectionLen--;
+    section[sectionLen] = '\0';
+
+    word32 outLen = ((sectionLen*3+3)/4)+1;
+    byte* out = XMALLOC(outLen*sizeof(byte), NULL, DYNAMIC_TYPE_PKCS7);
+    if (!out) return NULL;
+    byte* outHead = out;
+    ret = Base64_Decode((const byte*)section, sectionLen, out, &outLen);
+    if (ret < 0) {
+        WOLFSSL_MSG("Error base64 decoding S/MIME message.\n");
+        return NULL;
+    }
+    pkcs7 = wolfSSL_d2i_PKCS7(NULL, (const unsigned char**)&out, outLen);
+
+
+    MIME_free_headers(allHeaders);
+    if (outHead) XFREE(outHead, NULL, DYNAMIC_TYPE_PKCS7);
+    if (section) XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+
+
+    return pkcs7;
+}
+
+int MIME_parse_headers(WOLFSSL_BIO* in, int inLen, MimeHeader** headers)
+{
+    MimeHeader *nextHeader = XMALLOC(sizeof(MimeHeader), NULL, DYNAMIC_TYPE_PKCS7);
+    MimeHeader *curHeader = NULL;
+    MimeParameter *nextParam = XMALLOC(sizeof(MimeParameter), NULL, DYNAMIC_TYPE_PKCS7);
+    word32 start = 0;
+    word32 end = 0;
+    char *nameAttr = NULL;
+    char *bodyVal = NULL;
+    MimeTypes mimeType = MIME_HEADER;
+    MimeStatus mimeStatus = MIME_NAMEATTRIB;
+    int ret = 0;
+    int pos = 0;
+    int lineLen = 0;
+    byte isEnd = 0;
+    char* lines = XMALLOC((inLen+1)*sizeof(char), NULL, DYNAMIC_TYPE_PKCS7);
+    char* curLine = lines;
+
+    if (!nextHeader || !nextParam || !lines) return MEMORY_E;
```

Comment:
Should we also be sanitizing "in" and "headers" arguments?

Reply:
Definitely should be sanitizing in, fixed.  Not sure on headers as it is controlled fully by this function, and currently is being passed in as NULL in some cases.